compiler/src/java_plugin/cpp/java_generator.cpp

```diff
@@ -938,6 +871,54 @@ static void PrintBindServiceMethod(const ServiceDescriptor* service,
   p->Print("}\n");
 }
 
+// TODO(zdapeng): remove PrintDeprecatedDocComment once all our customers complete v0.15 upgrade.
+static void PrintDeprecatedDocComment(const ServiceDescriptor* service,
+                                      map<string, string>* vars,
+                                      Printer* p) {
+  p->Print(
+      *vars,
+      "/**\n"
+      " * This can not be used any more since v0.15.\n"
+      " * If your code using earlier version of gRPC-java is breaking when upgrading to v0.15,\n"
+      " * the following are suggested:\n"
+      " * <ul>\n"
+      " *   <li> replace {@code extends/implements $service_name$}"
+      " with {@code extends $service_name$ImplBase};</li>\n"
+      " *   <li> replace usage of {@code $service_name$} with {@code $service_name$ImplBase};"
+      "</li>\n"
+      " *   <li> replace usage of {@code Abstract$service_name$}"
+      " with {@link $service_name$ImplBase};</li>\n"
+      " *   <li> replace"
+      " {@code serverBuilder.addService($service_class_name$.bindService(serviceImpl))}\n"
+      " *        with {@code serverBuilder.addService(serviceImpl)};</li>\n"
+      " *   <li> if you are mocking stubs using mockito, please do not mock them."
+      " See the documentation\n"
+      " *        on testing with gRPC-java;</li>\n"
+      " *   <li> replace {@code $service_name$BlockingClient}"
+      " with {@link $service_name$BlockingStub};</li>\n"
+      " *   <li> replace {@code $service_name$FutureClient}"
+      " with {@link $service_name$FutureStub}.</li>\n"
+      " * </ul>\n"
+      " */\n");
+}
+
+// TODO(zdapeng): remove PrintDeprecated once all our customers complete v0.15 upgrade.
+static void PrintDeprecated(const ServiceDescriptor* service,
+                            map<string, string>* vars,
+                            Printer* p) {
+  (*vars)["service_name"] = service->name();
+  PrintDeprecatedDocComment(service, vars, p);
+  p->Print(*vars, "@Deprecated public static final class $service_name$ {}\n\n");
```

Comment:
`@java.lang.Deprecated`, since there could be another `@Deprecated` annotation in the package this is generated into.